hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -1225,130 +1302,292 @@ private boolean innerRename(Path source, Path dest)
       }
     }
 
-    // If we have a MetadataStore, track deletions/creations.
-    Collection<Path> srcPaths = null;
-    List<PathMetadata> dstMetas = null;
-    if (hasMetadataStore()) {
-      srcPaths = new HashSet<>(); // srcPaths need fast look up before put
-      dstMetas = new ArrayList<>();
-    }
-    // TODO S3Guard HADOOP-13761: retries when source paths are not visible yet
+    // Validation completed: time to begin the operation.
```

Comment:
That I will do. It might also line me up better for when I open up the rename/3 operation which will always throw an exception on any invalid state (rather than return "false" with no explanation)

Reply:
done. Not doing any tests on the validation alone, as the contract tests are expected to generate the failure conditions, but it does help isolate the two stages